File: ios/BMEVideoPlayerManager.m


// BMEVideoPlayerManager.m
#import "BMEVideoPlayerManager.h"
#import <React/RCTBridge.h>
#import <React/RCTUIManager.h>
#import "BMEVideoPlayerView.h"

@implementation BMEVideoPlayerManager

RCT_EXPORT_MODULE(BMEVideoPlayer)

- (UIView *)view
{
  return [[BMEVideoPlayerView alloc] init];
}

// Props
RCT_EXPORT_VIEW_PROPERTY(source, NSString)
RCT_EXPORT_VIEW_PROPERTY(paused, BOOL)
RCT_EXPORT_VIEW_PROPERTY(muted, BOOL)
RCT_EXPORT_VIEW_PROPERTY(volume, float)
RCT_EXPORT_VIEW_PROPERTY(resizeMode, NSString)

// Events
RCT_EXPORT_VIEW_PROPERTY(onPlaybackStatus, RCTBubblingEventBlock)

// Commands (UIManager dispatch)
RCT_EXPORT_METHOD(play:(nonnull NSNumber *)reactTag)
{
  [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, UIView *> *viewRegistry) {
    BMEVideoPlayerView *view = (BMEVideoPlayerView *)viewRegistry[reactTag];
    if (![view isKindOfClass:[BMEVideoPlayerView class]]) {
      RCTLogError(@"Cannot find BMEVideoPlayerView with tag #%@", reactTag);
      return;
    }
    [view play];
  }];
}

RCT_EXPORT_METHOD(pause:(nonnull NSNumber *)reactTag)
{
  [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, UIView *> *viewRegistry) {
    BMEVideoPlayerView *view = (BMEVideoPlayerView *)viewRegistry[reactTag];
    [view pause];
  }];
}

RCT_EXPORT_METHOD(seekTo:(nonnull NSNumber *)reactTag seconds:(double)seconds)
{
  [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, UIView *> *viewRegistry) {
    BMEVideoPlayerView *view = (BMEVideoPlayerView *)viewRegistry[reactTag];
    [view seekTo:CMTimeMakeWithSeconds(seconds, NSEC_PER_SEC) completion:nil];
  }];
}

RCT_EXPORT_METHOD(presentFullscreen:(nonnull NSNumber *)reactTag)
{
  [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, UIView *> *viewRegistry) {
    BMEVideoPlayerView *view = (BMEVideoPlayerView *)viewRegistry[reactTag];
    [view presentFullscreen];
  }];
}

@end

File: ios/ProjectName/bmevideo/BMEVideoPlayerManager.swift
import Foundation
import React

@objc(BMEVideoPlayerManager)
class BMEVideoPlayerManager: RCTViewManager {
  override func view() -> UIView! {
    return BMEVideoPlayerView()
  }

  override static func requiresMainQueueSetup() -> Bool {
    return true
  }

  @objc override func constantsToExport() -> [AnyHashable : Any]! {
    return ["Events": [
      BMEVideoPlayerEvent.onLoad,
      BMEVideoPlayerEvent.onProgress,
      BMEVideoPlayerEvent.onBuffer,
      BMEVideoPlayerEvent.onError,
      BMEVideoPlayerEvent.onEnd
    ]]
  }

  // MARK: - Props
  @objc func setSource(_ reactTag: NSNumber, src: NSDictionary) {
    bridge.uiManager.addUIBlock { (_, viewRegistry) in
      (viewRegistry?[reactTag] as? BMEVideoPlayerView)?.setSource(src)
    }
  }

  @objc func setPaused(_ reactTag: NSNumber, paused: Bool) {
    bridge.uiManager.addUIBlock { (_, viewRegistry) in
      (viewRegistry?[reactTag] as? BMEVideoPlayerView)?.setPaused(paused)
    }
  }

  @objc func setRepeat(_ reactTag: NSNumber, repeatVal: Bool) {
    bridge.uiManager.addUIBlock { (_, viewRegistry) in
      (viewRegistry?[reactTag] as? BMEVideoPlayerView)?.setRepeat(repeatVal)
    }
  }

  @objc func setMuted(_ reactTag: NSNumber, muted: Bool) {
    bridge.uiManager.addUIBlock { (_, viewRegistry) in
      (viewRegistry?[reactTag] as? BMEVideoPlayerView)?.setMuted(muted)
    }
  }

  @objc func setVolume(_ reactTag: NSNumber, volume: NSNumber) {
    bridge.uiManager.addUIBlock { (_, viewRegistry) in
      (viewRegistry?[reactTag] as? BMEVideoPlayerView)?.setVolume(volume)
    }
  }

  @objc func setRate(_ reactTag: NSNumber, rate: NSNumber) {
    bridge.uiManager.addUIBlock { (_, viewRegistry) in
      (viewRegistry?[reactTag] as? BMEVideoPlayerView)?.setRate(rate)
    }
  }

  // MARK: - Commands
  @objc func play(_ reactTag: NSNumber) {
    bridge.uiManager.addUIBlock { (_, viewRegistry) in
      (viewRegistry?[reactTag] as? BMEVideoPlayerView)?.play()
    }
  }

  @objc func pause(_ reactTag: NSNumber) {
    bridge.uiManager.addUIBlock { (_, viewRegistry) in
      (viewRegistry?[reactTag] as? BMEVideoPlayerView)?.pause()
    }
  }

  @objc func seekTo(_ reactTag: NSNumber, seconds: NSNumber) {
    bridge.uiManager.addUIBlock { (_, viewRegistry) in
      (viewRegistry?[reactTag] as? BMEVideoPlayerView)?.seekTo(seconds)
    }
  }

  @objc func releasePlayer(_ reactTag: NSNumber) {
    bridge.uiManager.addUIBlock { (_, viewRegistry) in
      (viewRegistry?[reactTag] as? BMEVideoPlayerView)?.releasePlayer()
    }
  }

  // MARK: - Preload
  @objc func preload(_ url: NSString) {
    PreloadManager.shared.preload(url as String)
  }

  @objc func cancelPreload(_ url: NSString) {
    PreloadManager.shared.cancel(url as String)
  }
}
